src/sgame/sg_buildable.cpp

```diff
@@ -1702,110 +1702,97 @@ itemBuildError_t G_CanBuild( gentity_t *ent, buildable_t buildable, int /*distan
 }
 
 /** Sets shared buildable entity parameters. */
-#define BUILDABLE_ENTITY_SET_PARAMS(params)\
-	params.oldEnt = ent;\
+template <typename entityType>
+void BUILDABLE_ENTITY_CREATE( gentity_t *ent, buildable_t buildable )
+{
+	typename entityType::Params params;
+	params.oldEnt = ent;
 	params.Health_maxHealth = BG_Buildable(buildable)->health;
+	ent->entity = new entityType(params);
+}
 
-#define BUILDABLE_ENTITY_START(entityType)\
-	entityType::Params params;\
-	BUILDABLE_ENTITY_SET_PARAMS(params);
-
-#define BUILDABLE_ENTITY_END(entityType)\
+template < typename entityType, typename std::enable_if<
+	std::is_same<entityType,DrillEntity>::value ||
+	std::is_same<entityType,LeechEntity>::value
+	>::type = 0>
+void BUILDABLE_ENTITY_CREATE( gentity_t *ent, buildable_t buildable )
```

Comment:
This seems overall less readable than what we have now. The specialization requires some verbose code duplication and is far from its point of use.

If you insist on replacing the macros with something else, it seems better to make template specializations for only the part that's different (i.e. nothing for most of them and the one line for mining structures). 

Personally if I were to simplify the original code somehow, I would just get rid of BUILDABLE_ENTITY_CREATE and write out BUILDABLE_ENTITY_START and BUILDABLE_ENTITY_END for each structure so that it's obvious which parts are in common.

Reply:
> it seems better to make template specializations for only the part that's different (i.e. nothing for most of them and the one line for mining structures).

I don't see how this would work. To have 2 template functions, as with before but instead of macros?

I've split the commit in 2, so that it would be easy to remove it since it's possibly how this will end.

Reply:
I'm not too sure how to do that, nor when i'll have time and motivation to do that, so I'll just remove this part from the PR. Easier for everyone like this (and it's not important at all anyway).